### ISSUE TITLE
Naming conventions

### ISSUE THREAD

**Issue by foolip** [web-platform-dx/web-features#24]
Having added a bunch of features (30 in total after https://github.com/web-platform-dx/feature-set/pull/23) I have a few questions:
- When should we use "css-" as a prefix?
  - css-nesting or just nesting? What do we do if we later need to disambiguate?
  - leading-trim or css-leading-trim?
- When should we use "-api" or "-element" as a suffix?
  - navigation-api, dialog vs. wbr-element in caniuse)
- When do we use the syntax vs. other well-known names for a feature? 
  - position-sticky after `position: sticky` vs. sticky-positioning after "Stick positioning"?
  - idle-detector after the interface name vs. idle-detection after the spec name?
  - container-queries after the feature name vs. at-container after `@container`?
- How do we name a feature that will grow, but is nevertheless known under its "larger" name?
  - container-queries is like this, the first shipping part is size container queries, but we'll soon have both size and style container queries. What does support for "container queries" mean then?

A related issue is how we can fix mistakes, so that we don't need to be as afraid of making them. I'm imagining:
- Aliases as a way to allow renaming, but clients can keep using the old names.
- Deprectation over a few major releases as a way to free up a name we need to use for something else.

**Comment by ddbeck** [comment]
> When should we use "css-" as a prefix?

This came up already on #23. I think plausible answers are:

1. Never
2. Always
3. Only when needed to resolve a naming conflict (or, alternatively, to use a suffix)

I'm partial to option 1 or the suffix variant on option 3. I think we should avoid baking technology domain divisions in feature identifiers, since I expect features to cross domains (e.g., a new CSS at-rule is accompanied by a new DOM API). By using prefixes, we might imply structural constraints on features that may not be true.

I think this extends to most everything else: call something by a reasonable name (with the caveat that we'll need to adapt over time if we learn that web developers tend toward some other name) and use suffixes to resolve name conflicts or real-world ambiguities.

More commentary on how this relates to some of the other questions soon.

**Comment by ddbeck** [comment]
More thoughts, mostly from the continued line commentary on #23.

> When should we use "-api" or "-element" as a suffix? 

It's interesting that the specs sometimes use "API" in the specification title (insert digression here on the ramifications of specs being named after the outcome not the concern). It seems like some things are going to be reasonably known as "Example API" or by a brand name (e.g., "WebRTC", "WebSockets") as opposed to common nouns (e.g., "web components", "cascade layers"). I don't think we can or should fight against this kind of thing, where it exists.

As a guideline, we might want to favor names to known to be used by web developers (and perhaps spec authors) over names we invent ourselves, even if it means we forgo some consistency. That is to say, we ought to describe already is, if there's already a definitive winner. That will sometimes lead us to "something-api"—oh well.

In cases where there's no conventional name yet, I'd like us to prefer common nouns and to avoid repetition and abbreviations. So given an API to control flying drones, we'd choose, in decreasing order of preference, something like: "drones", "drone-interface", "drone-api". I'd beg the spec authors to not force us into "webdrone".

**Comment by boazsender** [comment]
TLDR I don't think we should prefix/suffix grouping definitions because many features will probably cross the namespace boundary.

Longer version:

**On the one hand**, if we name-space groupings I think we'd want to scope CSS and APIs to their own name spaces with a directory. I also think we'd want to then add top level namespaces for JavaScript language features, and html features, which I'm sure you're also already thinking about. I imagine this could look something like:

* `feature-group-definitions/css/nesting.yml
* `feature-group-definitions/html/video.yml
* `feature-group-definitions/js/arrays.yml
* `feature-group-definitions/api/navigation.yml

I think the JS/API distinction here is especially useful because it helps a web developer reason about the difference between the JS language and the web platform api surface. It would be interesting to work with Node, Deno, and the other ones on share schema for this so someone could more easily smell the difference between the platforms.

**On the other hand**, I think I heard from the few conversations I've been in about this that the purpose of this work is to summarize features in web developer language, and often times this means crossing CSS/API/HTML/JS boundaries. "Video" might want to include the HTML element, the web APIs you can use to control it, the CSS you can use to style it, and etc. It might also want to link to other media APIs of interest that can be used in combination with it.  For this reason, I would stay away from scoping, or prefixing/suffixing.

Side note, is there a conversation happening about APIs existing in multiple feature groupings? Seems like that is how this will have to work. You're probably already thinking about this, too, I'm sure.

I might be lacking some context, so please let me know if I'm missing something, or talking about something out of scope.

**Comment by una** [comment]
I would highly recommend using feature names over syntax. I.e. using `container-queries` over `at-container` or `@container` because container queries encompass more than just that at-rule. For example, container query units are a part of the container query spec.

In regards to using a `css-`+ prefixing, many features span CSS and JS (or CSS and HTML, etc.). For example, scroll-driven animations have a CSS API and a JS API. Do you want to reference them as one feature or separately?

**Comment by foolip** [comment]
@boazsender 

I agree with your TLDR.

> Side note, is there a conversation happening about APIs existing in multiple feature groupings? Seems like that is how this will have to work. You're probably already thinking about this, too, I'm sure.

Do you have an example of the typical situation you're thinking about? What comes to mind for me is how many features have an [`allow` attribute value](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#attr-iframe-allow), but I can't think of any cases where the exact same API surface is part of two features. I'm imagining that most such cases amount to a dependency, where two features depend on another, but perhaps there are cases where it makes more sense to consider something part of two features independently rather than a dependency.

@una I agree that `at-container` in particular is not a good name. Over in https://github.com/web-platform-dx/feature-set/pull/23#issuecomment-1368668218 I also pondered the naming for Container Queries in particular, perhaps you want to weigh in there? In particular, what should we do about size container queries vs. style container queries vs. others that might come in the future?

**Comment by ddbeck** [comment]
@una

> In regards to using a `css-`+ prefixing, many features span CSS and JS (or CSS and HTML, etc.). For example, scroll-driven animations have a CSS API and a JS API. Do you want to reference them as one feature or separately?

I expect that features will span technologies like your example. I don't think they're reasonably separable as identifying things that web developers can use or do (e.g., if you've got some CSS feature with a corresponding interface, there's no sense in describing the interface separately, since it's useless without the CSS support itself).

I'm sure we'll get into the weeds where there will be debates about the separability of this or that feature. But, in general, we should expect some sound crossing.

**Comment by ddbeck** [comment]
I opened #38 to give form to some of the things we've discussed so far in this issue. It's by no means a complete answer to the questions here, but I did want to seed the practice of producing documentation out of discussions like this one.

**Comment by boazsender** [comment]
@foolip 
> Do you have an example of the typical situation you're thinking about?

I was thinking about how media streams could be considered part of video, audio and canvas when I wrote this. I'm drawing a blank on other examples, but will continue to think about it.

**Comment by foolip** [comment]
@boazsender `MediaStream` is a good example. It's connected to `getUserMedia()`, media elements, WebRTC, Web Audio and perhaps more. But it's not really a usable feature on its own, it's just an empty pipe connecting other features.

But because the `MediaStream` API has evolved over time, I think it makes sense to define it as its own feature, and make others depend on it.

**Comment by foolip** [comment]
This is now covered by https://github.com/web-platform-dx/web-features/tree/main/docs.
